File: bindings/java/samples/Sample_sparc.java
/*

Java bindings for the Unicorn Emulator Engine

Copyright(c) 2015 Chris Eagle

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/

/* Unicorn Emulator Engine */
/* By Nguyen Anh Quynh, 2015 */

/* Sample code to demonstrate how to emulate Sparc code */

import unicorn.*;

public class Sample_sparc {

   // code to be emulated
   public static final byte[] SPARC_CODE = {-122,0,64,2};
   //public static final byte[] SPARC_CODE = {-69,112,0,0}; //illegal code
   
   // memory address where emulation starts
   public static final int ADDRESS = 0x10000;
   
   public static final long toInt(byte val[]) {
      long res = 0;
      for (int i = 0; i < val.length; i++) {
         long v = val[i] & 0xff;
         res = res + (v << (i * 8));
      }
      return res;
   }

   public static final byte[] toBytes(long val) {
      byte[] res = new byte[8];
      for (int i = 0; i < 8; i++) {
         res[i] = (byte)(val & 0xff);
         val >>>= 8;
      }
      return res;
   }

   // callback for tracing basic blocks
   private static class MyBlockHook implements BlockHook {
      public void hook(Unicorn u, long address, int size, Object user_data) {
         System.out.print(String.format(">>> Tracing basic block at 0x%x, block size = 0x%x\n", address, size));
      }
   }
      
   // callback for tracing instruction
   private static class MyCodeHook implements CodeHook {
      public void hook(Unicorn u, long address, int size, Object user_data) {
         System.out.print(String.format(">>> Tracing instruction at 0x%x, instruction size = 0x%x\n", address, size));
      }
   }
      
   static void test_sparc()
   {
       byte[] g1 = toBytes(0x1230);     // G1 register
       byte[] g2 = toBytes(0x6789);     // G2 register
       byte[] g3 = toBytes(0x5555);     // G3 register
   
       System.out.print("Emulate SPARC code\n");
   
       // Initialize emulator in Sparc mode
       Unicorn u = new Unicorn(Unicorn.UC_ARCH_SPARC, Unicorn.UC_MODE_32 + Unicorn.UC_MODE_BIG_ENDIAN);
   
       // map 2MB memory for this emulation
       u.mem_map(ADDRESS, 2 * 1024 * 1024, Unicorn.UC_PROT_ALL);
   
       // write machine code to be emulated to memory
       u.mem_write(ADDRESS, SPARC_CODE);
   
       // initialize machine registers
       u.reg_write(Unicorn.UC_SPARC_REG_G1, g1);
       u.reg_write(Unicorn.UC_SPARC_REG_G2, g2);
       u.reg_write(Unicorn.UC_SPARC_REG_G3, g3);
   
       // tracing all basic blocks with customized callback
       u.hook_add(new MyBlockHook(), 1, 0, null);
   
       // tracing one instruction at ADDRESS with customized callback
       u.hook_add(new MyCodeHook(), ADDRESS, ADDRESS, null);

       // emulate machine code in infinite time (last param = 0), or when
       // finishing all the code.
       u.emu_start(ADDRESS, ADDRESS + SPARC_CODE.length, 0, 0);
   
       // now print out some registers
       System.out.print(">>> Emulation done. Below is the CPU context\n");
   
       g3 = u.reg_read(Unicorn.UC_SPARC_REG_G3, 4);
       System.out.print(String.format(">>> G3 = 0x%x\n", toInt(g3)));
   
       u.close();
   }
   
   public static void main(String args[])
   {
      test_sparc();
   }
}

File: bindings/haskell/samples/SampleSparc.hs
-- Sample code to demonstrate how to emulate Sparc code

import Unicorn
import Unicorn.Hook
import qualified Unicorn.CPU.Sparc as Sparc

import qualified Data.ByteString as BS
import Data.Word
import qualified Numeric as N (showHex)

-- Code to be emulated
--
-- add %g1, %g2, %g3
sparcCode :: BS.ByteString
sparcCode = BS.pack [0x86, 0x00, 0x40, 0x02]

-- Memory address where emulation starts
address :: Word64
address = 0x10000

-- Pretty-print integral as hex
showHex :: (Integral a, Show a) => a -> String
showHex =
    flip N.showHex ""

-- Calculate code length
codeLength :: Num a => BS.ByteString -> a
codeLength =
    fromIntegral . BS.length

hookBlock :: BlockHook ()
hookBlock _ addr size _ =
    putStrLn $ ">>> Tracing basic block at 0x" ++ showHex addr ++
               ", block size = 0x" ++ (maybe "0" showHex size)

hookCode :: CodeHook ()
hookCode _ addr size _ =
    putStrLn $ ">>> Tracing instruction at 0x" ++ showHex addr ++
               ", instruction size = 0x" ++ (maybe "0" showHex size)

testSparc :: IO ()
testSparc = do
    putStrLn "Emulate SPARC code"

    result <- runEmulator $ do
        -- Initialize emulator in Sparc mode
        uc <- open ArchSparc [ModeSparc32, ModeBigEndian]

        -- Map 2MB memory for this emulation
        memMap uc address (2 * 1024 * 1024) [ProtAll]

        -- Write machine code to be emulated to memory
        memWrite uc address sparcCode

        -- Initialize machine registers
        regWrite uc Sparc.G1 0x1230
        regWrite uc Sparc.G2 0x6789
        regWrite uc Sparc.G3 0x5555

        -- Tracing all basic blocks with customized callback
        blockHookAdd uc hookBlock () 1 0

        -- Tracing all instructions with customized callback
        codeHookAdd uc hookCode () 1 0

        -- Emulate machine code in infinite time (last param = Nothing), or
        -- when finishing all the code
        let codeLen = codeLength sparcCode
        start uc address (address + codeLen) Nothing Nothing

        -- Return results
        g3 <- regRead uc Sparc.G3

        return g3
    case result of
        Right g3 -> do
            -- Now print out some registers
            putStrLn ">>> Emulation done. Below is the CPU context"
            putStrLn $ ">>> G3 = 0x" ++ showHex g3
        Left err -> putStrLn $ "Failed with error: " ++ show err ++ " (" ++
                               strerror err ++ ")"

main :: IO ()
main =
    testSparc
